### PR TITLE
hack session that tries to encapsulate a project structure

### PR DESCRIPTION
**Pull request by cpdeethree**
This is VERY MUCH WIP. Nothing runs or works. It's only to have a discussion about some project structure and hack on with connector ops team. A lot of this is pretty stream of consciousness, but I wanted to get this out there so we can talk about how this might work as early as possible.

# What

We want to have reusable pipeline components that are shared across repositories. It's intended to be an OSS mirror of https://github.com/airbytehq/workflow-actions, our internal equivalent. Ideally anything OSS would go in here, and then if we needed some internal-only stuff (hopefully rare), it could go in that other repository

From a high level the idea is as follows

- Create a reusable set of pipeline components in this repository. This includes the models we use to define pipelines, reusable stuff that we want everywhere (auth to gcloud, auth to aws, upload to s3, file/folder utility functions)
- Publish everything not in _pipelines folder as a PyPi package from this repository

- Every repository (airbytehq/airbyte, airbytehq/airbyte-platform, airbytehq/airbyte-types, airbytehq/airbyte-protocol) has some folder, currently that i've called _pipelines, that they implement repository specific stuff. There's a pyproject.toml and poetry definition at top level of all of these that allows a developer to have a shared interface as they jump from repo to repo. Meaning that every developer can simply run `poetry install`, `poetry run _pipelines/ci.py`, `poetry run mypy _pipelines/ci.py` etc. to interface with the pipeline code. 
- We can lift this up into a click CLI like in connectors repo works now. Running individual arguments, options, etc can be passed via the command line via click, and those are also what workflow_dispatch takes as input in the github actions YML
-  Every repository brings in this airbyte_actions package as a dependency
- That _pipelines folder contains repo specific stuff in actions directroy
- at top level of _pipelines, there is always a ci.py that is the entrypoint for GHA inputs. This ci.py is the only thing ever invoked from ci.yml. There is only one ci.yml and ci.py in every repository. The ci.py handles:
  -  declaring pipelines
  - figuring out which pipelines are going to run and with what parameters based on files changed or arguments passed into ci.yml via dispatch
  - establishing a GlobalContext that is shared across pipeline runs (dagger config/client is an example that is passed around in here) 
  - taking in workflow arguments that are passed in from the github action (e.g. workflow_dispatch, if I only want to run a single pipeline or I have some specific pipeline parameters)
  - Implement an extendable GlobalTestReport that contains some standardized logging/reporting across top level runs
- Also in the top level of _pipelines, we have individual pipelines that are declared, like in this case actions_runner.py. These pipelines
  - Establish a pipeline context (auto tagging of labels in Dagger for visibility in UI)
  - Implement an extendable PipelineTestReport that contains some standardized logging/reporting across pipelines
  - reference container "environments" in actions that provide utilities for establishing, returning containers. These are almost always specific to a repository, and are not intended to be shared across repositories. This may cause a small amount of duplicated code, but removes the developer's overhead of deciding whether an "environment" should be abstracted away. We can always go back later and move some of these concepts into the shared airbyte_actions repository
  
  
 Why not put this in airbytehq/airbyte?
  
  Same reason as defined in https://docs.google.com/document/d/1jl122NkgH9rs6mr58UgyOVqD9J4d_656BcCEMlVVfbQ/edit#heading=h.hxf91ev4axxq
  
  basically, it's possible, but if the code is genuinely meant to be reused across repositories, like in this case, it muddies the dependency situation

### CODE REVIEW

File: models/pipelines.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import asyncer
+from utils.dagger import with_exit_code, with_stderr, with_stdout
+from dagger import Client, Container
+from rich.style import Style
+
+if TYPE_CHECKING:
+    from contexts import GlobalContext
+
+
+class StepStatus(Enum):
```
Comment (cpdeethree) [inline models/pipelines.py:22]: This is a neat way of unifying pipeline components. I suspect some abstraction like this was purposefully left out of the dagger model in order to see what end users implemented, but maybe eventually makes sense to go into the dagger model at some point in the future

File: models/contexts.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        return True
+
+
+class PipelineContext(GlobalContext):
+    pass
+    """The pipeline context is used to store configuration for a specific pipeline run. 
+    
+    It extends the GlobalContext class"""
+
+class GitContext(GlobalContext):
```
Comment (cpdeethree) [inline models/contexts.py:158]: I'm not sure whether GitContext is necessary, and I am also not sure whether it should extend GlobalContext. I thought about creating a model to abstract away all of the git related behavior to make it easy for pipelines to ask questions about file-structure related stuff. But it also feels like a premature abstraction

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      - name: Install Python 3.10
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
+      - name: Install pipeline dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install poetry
+          poetry install
+      - name: Run actions pipeline [WORKFLOW DISPATCH]
```
Comment (cpdeethree) [inline .github/workflows/ci.yaml:38]: This is a simple copy paste, but in the proposed structure we would only have one step to run actions pipeline, and this behavior is delegated. Since we have EKS runners now, and when python is baked in, the quick step of running dagger to determine whether sub-pipelines are invoked every single time is probably going to be quite acceptable in terms of speed.

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+          DAGGER_TMP_BINDIR="/tmp/dagger_${DAGGER_CLI_COMMIT}"
+          export _EXPERIMENTAL_DAGGER_CLI_BIN="$DAGGER_TMP_BINDIR/dagger"
+          if [ ! -f  "$_EXPERIMENTAL_DAGGER_CLI_BIN" ]; then
+              mkdir -p "$DAGGER_TMP_BINDIR"
+              curl "https://dl.dagger.io/dagger/main/${DAGGER_CLI_COMMIT}/dagger_${DAGGER_CLI_COMMIT}_$(uname -s | tr A-Z a-z)_$(uname -m | sed s/x86_64/amd64/).tar.gz" | tar xvz -C "$DAGGER_TMP_BINDIR"
+          fi
+          poetry run python .pipelines/ci.py ${{ github.event.inputs.test-connectors-options }}
+        env:
+          GCP_GSM_CREDENTIALS: ${{ secrets.GCP_GSM_CREDENTIALS }}
+          CI_GITHUB_ACCESS_TOKEN: ${{ secrets.GH_PAT_MAINTENANCE_OCTAVIA }}
```
Comment (cpdeethree) [inline .github/workflows/ci.yaml:71]: these CI specific variables, I think, should also be pulled in the dagger code rather than in the YML. I think it belongs in a Github CI model and function that maps the Github CI specific environment variables to generic environment variables that we declare. If we did this, it would be trivial if we ran in jenkins or somewhere else to just make a jenkins model that did similar mapping and just swap out that model that maps to our generic environment variables.

 I think the only thing we will HAVE to use here is GCP_GSM_CREDENTIALS (in the secrets v2 tech spec that is currently being circulated, you need these credentials to access all other secrets)

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,75 @@
+name: Dagger CI
+
+on:
+  workflow_dispatch:
+    inputs:
+      pipeline-options:
```
Comment (cpdeethree) [inline .github/workflows/ci.yaml:6]: I like this. I do fear in the model I'm proposing that this might become overloaded with a ton of options in our larger repositories once we migrate more pipelines to dagger. I'm not sure what the correct solution is yet, but it kind of sucks that we are limited in a way with what we can pass to workflow_dispatch (max 10 parameters are available in the GUI).

File: pyproject.toml [hunk 6]
```diff
@@ -2,18 +2,24 @@
...
 
 [tool.poetry.dependencies]
 python = "^3.10"
-dagger-io = "^0.4.0"
+dagger-io = "^0.4.2"
 python-dotenv = "^0.21.0"
 asyncclick = "^8.1.3.4"
 google-cloud-run = "^0.5.0"
 pre-commit = "^2.21.0"
+mypy = "^1.1.1"
```
Comment (cpdeethree) [inline pyproject.toml:15]: As discussed in my chat with augustin, i think mypy is important in the context of dagger code because the lazy evaluation model dictates the behavior of dagger based on the types you are working with. So strict type enforcement forces the developer to be fully aware of what they are working on. I think the value outside of dagger code is a bit more questionable (e.g. if we wanted mypy on connectors code, would the strict typing actually discourage connector development?)

File: remote_storage/s3.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+from pathlib import Path
+
+from ..utils.dagger import with_exit_code
+from dagger import Client, File, Secret
+
+
+async def upload_to_s3(dagger_client: Client, file_to_upload_path: Path, key: str, bucket: str) -> int:
```
Comment (cpdeethree) [inline remote_storage/s3.py:10]: A very good example of the kind of code that we want to share across many repositories

File: reporting/test_report.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                sub_panels.append(sub_panel)
+            failures_group = Group(*sub_panels)
+            to_render.append(failures_group)
+
+        main_panel = Panel(Group(*to_render), title=main_panel_title, subtitle=duration_subtitle)
+        console.print(main_panel)
+
+
+@dataclass(frozen=True)
+class PipelineTestReport:
```
Comment (cpdeethree) [inline reporting/test_report.py:110]: This is just a quick and not-cleaned-up copy paste, but I think PipelineTestReport should extend either GlobalTestReport or they should have a base/abstract class called TestReport. Maybe the same for Contexts as well

File: utils/dagger.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
+#
```
Comment (cpdeethree) [inline utils/dagger.py:1]: Should probably rename this file to dagger_helpers to avoid naming conflict confusion since the dagger package is called dagger

File: utils/repo.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def get_modified_files(current_git_branch: str, current_git_revision: str, diffed_branch: str, is_local: bool = True) -> Set[str]:
+    if is_local:
+        return get_modified_files_local(current_git_revision, diffed_branch)
+    else:
+        return anyio.run(get_modified_files_remote, current_git_branch, current_git_revision, diffed_branch)
+    
+
+def is_local() -> bool:
```
Comment (cpdeethree) [inline utils/repo.py:73]: I think that since this references an arbitrary github specific env variable, as discussed before, this needs to go in its own file that maps these. Either that or the github specific environment variable needs to be mapped to same name generic environment variable that the pipeline code then consumes

File: _pipelines/ci.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+                               , git_branch = get_current_git_branch()
+                               , git_revision = get_current_git_revision()
+                               , git_repo_url = GIT_REPOSITORY)
+
+logging.basicConfig(level=logging.INFO, format="%(name)s: %(message)s", datefmt="[%X]", handlers=[RichHandler(rich_tracebacks=True)])
+
+logger = logging.getLogger(__name__)
+
+# PASS IN CONFIG VIA CLI 
+async def run(context: GlobalContext, **kwargs) -> GlobalTestReport:
```
Comment (alafanechere) [inline _pipelines/ci.py:29]: @bnchrch made a great suggestion in another PR: we think the "what to run" logic should be attached to GlobalContext/PipelineContext subclasses. 
The entrypoint could be a shared one for all projects using the global package:
```python
async def run_pipeline(context) -> Report:
    async with context:
        report = await context.run()
    return report

async def run_all_pipelines(contexts: List[PipelineContext]): 
    async with asyncer.create_task_group() as tg:
        tasks = [tg.soonify(run_pipeline(context)) for context in contexts]
    return GlobalTestReport(reports=[t.value for t in tasks])
```

Then all the CLI roles we'd be implementing in other repo would be to instantiate PipelineContext object according to user inputs and pass a list of pipelines to `run_all_pipelines`.

File: reporting/test_report.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+from rich.panel import Panel
+from rich.style import Style
+from rich.table import Table
+from rich.text import Text
+
+console = Console()
+
+
+@dataclass(frozen=True)
+class GlobalTestReport:
```
Comment (alafanechere) [inline reporting/test_report.py:29]:
```suggestion
class GlobalReport:
```
I think we should have two kind of reports: pipeline specific ones that will provide details about a pipeline execution, and a global one that would aggregate the reports of all the pipelines that got executed in the global context. I think `GlobalContext` and `GlobalReport` should be singletons. `GlobalContext` holds all the pipelines contexts that need to run, GlobalReport gather the execution results, shared through pipeline specific reports.

File: models/pipelines.py
Comment (alafanechere) [inline models/pipelines.py:1]: Can we rename this module to `steps` as it declares step related classes?

File: models/pipelines.pyi
Comment (alafanechere) [inline models/pipelines.pyi:1]: Is this autogenerated?

File: _pipelines/ci.py (hunk 11 above)
Comment (cpdeethree) [reply _pipelines/ci.py:29]: Interesting! I want to see what this would look like in practice. I think passing a list of pipelines via CLI makes sense, I am just curious how we can solve the problem of mapping all of the various possible options for each pipeline that one might want to pass that isn't ugly

For instance, I am very afraid under the single entrypoint model that we end up with something like:

```bash
poetry run python --pipelines=connectors_ci, image_publish --pipeline_options={connectors_ci: connectors=source-pokeapi, image_publish: publish=false}
```

Which is of course nearly defeating the point of having it in a CLI in the first place ;)

File: models/pipelines.py
Comment (cpdeethree) [reply models/pipelines.py:1]: Yes!

File: models/pipelines.pyi
Comment (cpdeethree) [reply models/pipelines.pyi:1]: it is in fact generated with the `stubgen` command. It helps mypy with type hinting. https://mypy.readthedocs.io/en/stable/stubs.html

I'm not sure it's actually necessary in this use-case yet, just added it as an example

File: _pipelines/ci.py (hunk 11 above)
Comment (alafanechere) [reply _pipelines/ci.py:29]: I meant that each project could implement the CLI they like to match their use case, but they all end up calling `run_all_pipelines`.
I'd prefer to not enforce any CLI structure (or maybe a global command group) in this repo to give maximal flexibility on implementation. We enforce the interface through the use of (abstract) classes, not through the CLI.

File: src/models/pipeline.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    title: ClassVar
+
+    def __init__(self, context: GlobalContext) -> None:
+        self.global_context = context
+        self.pipeline_context = PipelineContext(
+            is_local=context.is_local,
+            git_branch=context.git_branch,
+            git_revision=context.git_revision,
+            git_repo_url=context.git_repo_url,
+        )
```
Comment (alafanechere) [inline src/models/pipeline.py:58]:
```suggestion

    def __init__(self, context: GlobalContext) -> None:
        self.global_context = context
        dagger_client = self.global_context.dagger_client.pipeline(self.title)
        self.pipeline_context = PipelineContext.with_global_context(global_context).with_dagger_client(dagger_client)
```

File: src/models/pipeline.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+    @abstractmethod
+    async def _run(self) -> PipelineResult:
```
[the remark this answers is not in the document]
Comment (cpdeethree) [reply src/models/pipeline.py:61]: Yes. we need a way to bail out when a specific pipeline fails. More generically we need a way to declare dependencies between tasks
